ipn/ipnauth/access_check.go

```diff
@@ -0,0 +1,346 @@
+func FilterProfile(caller Identity, profile ipn.LoginProfileView, prefs ipn.PrefsGetter) ipn.LoginProfileView {
+	switch {
+	case CheckProfileAccess(caller, profile, prefs, ReadProfileInfo).Allowed():
+		return profile
+	default:
+		res := &ipn.LoginProfile{
+			ID:             profile.ID(),
+			Key:            profile.Key(),
+			LocalUserID:    profile.LocalUserID(),
+			UserProfile:    maskedUserProfile(profile),
+			NetworkProfile: maskedNetworkProfile(profile),
+		}
+		res.Name = res.UserProfile.LoginName
+		return res.View()
+	}
+}
+
+// maskedNetworkProfile returns a masked tailcfg.UserProfile for the specified profile.
+func maskedUserProfile(ipn.LoginProfileView) tailcfg.UserProfile {
+	return tailcfg.UserProfile{
+		LoginName:     "Other User's Account",
+		DisplayName:   "Other User",
+		ProfilePicURL: "",
+	}
+}
+
+// maskedNetworkProfile returns a masked ipn.NetworkProfile for the specified profile.
+func maskedNetworkProfile(ipn.LoginProfileView) ipn.NetworkProfile {
+	return ipn.NetworkProfile{
+		DomainName: "",
+	}
+}
+
+// PrefsChangeRequiredAccess returns access required to change prefs as requested by mp.
+func PrefsChangeRequiredAccess(mp *ipn.MaskedPrefs) ProfileAccess {
+	masked := reflect.ValueOf(mp).Elem()
+	return maskedPrefsFieldsAccess(&mp.Prefs, "", masked)
+}
+
+// maskedPrefsFieldsAccess returns access required to change preferences, whose
+// corresponding {FieldName}Set flags are set in masked, to the values specified in p.
+// Path represents a dot-separated path to masked from the ipn.MaskedPrefs root.
+func maskedPrefsFieldsAccess(p *ipn.Prefs, path string, masked reflect.Value) ProfileAccess {
+	var access ProfileAccess
+	for i := 0; i < masked.NumField(); i++ {
+		fName := masked.Type().Field(i).Name
+		if !strings.HasSuffix(fName, "Set") {
+			continue
+		}
+		fName = strings.TrimSuffix(fName, "Set")
+		fPath := path + fName
+		fValue := masked.Field(i)
+
+		switch fKind := fValue.Kind(); fKind {
+		case reflect.Bool:
+			if fValue.Bool() {
+				access |= prefsFieldRequiredAccess(p, fPath)
+			}
+		case reflect.Struct:
+			access |= maskedPrefsFieldsAccess(p, fPath+".", fValue)
+		default:
+			panic(fmt.Sprintf("unsupported mask field kind %v", fKind))
+		}
+	}
+	return access
+}
+
+// prefsDefaultFieldAccess is the default ProfileAccess required to modify ipn.Prefs fields
+// that do not have access rights overrides.
+const prefsDefaultFieldAccess = ChangePrefs
+
+var (
+	// prefsStaticFieldAccessOverride allows to override ProfileAccess needed to modify ipn.Prefs fields.
+	// The map uses dot-separated field paths as keys.
+	prefsStaticFieldAccessOverride = map[string]ProfileAccess{
+		"ExitNodeID":             ChangeExitNode,
+		"ExitNodeIP":             ChangeExitNode,
+		"ExitNodeAllowLANAccess": ChangeExitNode,
+		"LoggedOut":              DeleteProfile, // Logging out should require the same level of access as profile deletion.
```

Comment:
maybe? maybe not? dunno

Reply:
It appears clients never set this value via `EditPrefs` and instead use `Logout`.

It also appears that if a client _were_ to set `LoggedOut` via `EditPrefs`, that wouldn't result in profile deletion.
But calling `LocalBackend.Logout` [would](https://github.com/tailscale/tailscale/blob/03d5d1f0f951b7964f2aa2c74ea395e511aff605/ipn/ipnlocal/local.go#L4637) .

I personally find it a little confusing, and perhaps `ipn.Prefs` isn't the best place for the `LoggedOut` _state_.
But I decided to remove that line for the time being. Thank for flagging!